### PR TITLE
feat(mojaloop/3471): profile cl position handler

### PR DESCRIPTION
**Pull request by vijayg10**
feat(mojaloop/3471): profile cl position handler

### CODE REVIEW

**Comment by kleyow** [comment]
The kafka dashboards are empty, is that expected?

**Comment by vijayg10** [comment]
> The kafka dashboards are empty, is that expected?

Thanks @kleyow , not sure what happened with those dashboards. I just look at the central ledger related dashboards where we can get kafka message lag also. So I am just removing those empty dashboards for now.

File: cpu-profiling/cl-position-handler/README.md [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
...ollows:** <br> 1. No considerable issues detected in profiling with Chrome debugger <br>2. No issues detected  with Clinic.js bubbleprof analyser <br>3. If CACHE is disabled, there are `13` queries are being executed during processing a `action=prepare` message. And `8` queries during `action=commit` message including BEGIN and COMMITs.<br>4. For `5000` messages the total queries for `action=prepare` are `65000` if CACHE is disabled. This matches with the query count in single transfer (i.e. 5000x13)<br>5. If CACHE is enabled, for `5000` messages the total queries for `action=prepare` are reduced to `45000` 🔽<br>6. The indexes are properly set and no slow queries detected<br><br> **Observed areas that are limiting throughput and scalability:**<br> 1. `Batch processing` implementation is broken and can not be enabled which could improve performance <br> 2. With current setup (kafka key and partition combination), `Scalability` of position handler is not possible for a single DFSP<br><br> _Note: We are not able to determine any single cause of the bottleneck at this time, the load should be handled through batching and scalability._ | See [#follow-up-stores](#follow-up-stories) |   |
```
Comment (mdebarros) [inline cpu-profiling/cl-position-handler/README.md]: 1. Include the number of queries that occur when caching is enabled.
2. Include findings for "scalability" when more than now DFSP is being processed.
3. Include high-level findings of PRISM

File: cpu-profiling/cl-position-handler/README.md [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+  ```
+  By looking at the query log in mysql, the connections are being reused, so it seems like there is no issue.
+
+---
+
+## Mysql queries
+
+The following are the queries executed during a prepare / commit message handling operation.
+
+### Prepare
```
Comment (mdebarros) [inline cpu-profiling/cl-position-handler/README.md:132]: Can you also indicate what queries are cached when caching is enabled for both prepare and fulfil (i.e. commit) please.

File: cpu-profiling/cl-position-handler/README.md [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+2023-08-21T12:04:15.923564Z        41 Query     select * from `participantLimit`
+2023-08-21T12:04:15.924732Z        28 Query     update `participantPosition` set `value` = '2.0000', `reservedValue` = '0.0000', `changedDate` = '2023-08-21 12:04:15.921' where `participantPositionId` = 7
+2023-08-21T12:04:15.925035Z        28 Query     select * from `transfer` where `transferId` in ('9d74acdb-d7da-4362-8f45-661d4fb2c12e') for update
+2023-08-21T12:04:15.926852Z        28 Query     insert into `transferStateChange` (`createdDate`, `reason`, `transferId`, `transferStateChangeId`, `transferStateId`) values ('2023-08-17 15:25:07.000', NULL, '9d74acdb-d7da-4362-8f45-661d4fb2c12e', NULL, 'RESERVED')
+2023-08-21T12:04:15.928243Z        28 Query     insert into `participantPositionChange` (`participantPositionId`, `reservedValue`, `transferStateChangeId`, `value`) values (7, '0.0000', 5073, '2.0000')
+2023-08-21T12:04:15.928811Z        28 Query     COMMIT
+```
+
+### Commit
```
Comment (mdebarros) [inline cpu-profiling/cl-position-handler/README.md]:
```suggestion
### Commit (i.e. Fulfil)
```

**Review by mdebarros — CHANGES_REQUESTED** [review]
See comments.

File: cpu-profiling/cl-position-handler/README.md (hunk 3 above)
Comment (mdebarros) [reply cpu-profiling/cl-position-handler/README.md]: if You can capture the "smoothness" of the flow-mode vs recursive.

**Comment by mdebarros** [comment]
@vijayg10 some other "observations" that I think we should include:

1. The Prepare/Fulfil handlers are not correctly setting the optimised Message Key as the FSPIOP-Destination (DFSP-ID) is currently being used. This may result in unnecessary MySQL row lock contention. We need a follow-up story for this as this should be changed as follows:
    - Prepare should be using the PayerFSP-ID as the Message Key
    - Fulfil should be using the PayeeFSP-ID as the Message Key
    - Abort should be using the PayerFSP-ID as the Message Key
2. We can further increase the concurrent processing per DFSP by using the Account-ID instead of the DFSP-ID as the Message Key. <-- we should also create a follow-up story for this
3. Also document that the current "batch-process" logic of the position handler (and most likely other handlers, e.g. prepare, fulfil, and even notifications) is failing due to an error/bug that requires further investigation. (ref: https://github.com/mojaloop/project/issues/3488,  https://github.com/mojaloop/project/issues/3489)
4. We should also avoid using a hash partition strategy due to the extra partition overhead, and head-ache of partition management based on the key-space to ensure an even distribution. We should instead use a custom consistent partitioner strategy. <-- we should also create a follow-up story for this

**Comment by kleyow** [comment]
> > The kafka dashboards are empty, is that expected?
> 
> Thanks @kleyow , not sure what happened with those dashboards. I just look at the central ledger related dashboards where we can get kafka message lag also. So I am just removing those empty dashboards for now.

@vijayg10 Changing kafka-exporter to use 
```
    command:
      - --kafka.server=kafka:29092
```
should fix it

File: cpu-profiling/cl-position-handler/README.md (hunk 1 above)
Comment (vijayg10) [reply cpu-profiling/cl-position-handler/README.md]: Done

File: cpu-profiling/cl-position-handler/README.md (hunk 3 above)
Comment (vijayg10) [reply cpu-profiling/cl-position-handler/README.md]: Done

**Review by mdebarros — APPROVED** [review]

**Pull request merged**